runtime/lua/vim/_memoize.lua

```diff
@@ -0,0 +1,51 @@
+--- Module for private utility functions
+
+local function concat_hash(...)
+  return table.concat({...})
+end
+
+local function idx_hash(idx)
+  return function(...)
+    return select(idx, ...)
+  end
+end
+
+--- @param hash_fn any
+--- @return fun(...): any
+local function resolve_hash(hash_fn)
+  if type(hash_fn) == 'number' then
+    hash_fn = idx_hash(hash_fn)
+  elseif type(hash_fn) == 'string' then
+    if hash_fn == 'concat' then
+      hash_fn = concat_hash
+    else
+      error('invalid value for hash_fn: '..hash_fn)
+    end
+  end
+  return hash_fn
+end
+
+--- @generic F: function
+--- @param hash? integer|'concat'|fun(...): any
+--- @param fn F
+--- @return F
+return function(hash, fn)
+  vim.validate({
+    hash = { hash, {'integer', 'string', 'function' } },
+    fn = { fn, 'function' },
+  })
+
+  ---@type table<any,table<any,any>>
+  local cache = setmetatable({}, { __mode = 'kv' })
+
+  hash = resolve_hash(hash)
+
+  return function(...)
+    local key = hash(...)
+    if cache[key] == nil then
+      cache[key] = vim.F.pack_len(fn(...))
```

Comment:
Since this is not doing deepcopy anymore, maybe the docs should clarify that you will get a shared reference (so be careful about mutating)

Reply:
Sure. 

Btw, the reason I decided `deepcopy()` isn't worth it is that you can't protect against upvals being changed in closures, in addition to not being able to copy userdata.

Also the cache is invalidated fairly often, so mutation is _less_ likely to cause issues.